engine/gamesys/src/gamesys/components/comp_particlefx.cpp

```diff
@@ -254,48 +248,118 @@ namespace dmGameSystem
         return dmGameObject::UPDATE_RESULT_OK;
     }
 
+    static inline int32_t GetEmitterAttributeIndex(const dmGraphics::VertexAttribute* attributes, uint32_t attributes_count, dmhash_t name_hash)
+    {
+        for (int i = 0; i < attributes_count; ++i)
+        {
+            if (attributes[i].m_NameHash == name_hash)
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    static void FillEmitterAttributeInfos(const dmGraphics::VertexAttribute* attributes, uint32_t attributes_count, dmParticle::ParticleVertexAttributeInfos* infos)
+    {
+        for (int i = 0; i < infos->m_NumInfos; ++i)
+        {
+            int32_t emitter_attribute_index = GetEmitterAttributeIndex(attributes, attributes_count, infos->m_Infos[i].m_NameHash);
+            if (emitter_attribute_index >= 0)
+            {
+                infos->m_Infos[i].m_NameHash = attributes[emitter_attribute_index].m_NameHash;
+                dmGraphics::GetAttributeValues(attributes[emitter_attribute_index], (const uint8_t**) &infos->m_Infos[i].m_ValuePtr, &infos->m_Infos[i].m_ValueByteSize);
+            }
+        }
+    }
+
+    static void FillParticleMaterialAttributeInfos(dmRender::HMaterial material, dmParticle::ParticleVertexAttributeInfos* infos)
+    {
+        const dmGraphics::VertexAttribute* material_attributes;
+        uint32_t material_attributes_count;
+        dmRender::GetMaterialProgramAttributes(material, &material_attributes, &material_attributes_count);
+
+        infos->m_NumInfos     = dmMath::Min(material_attributes_count, (uint32_t) dmGraphics::MAX_VERTEX_STREAM_COUNT);
+        infos->m_VertexStride = dmGraphics::GetVertexDeclarationStride(dmRender::GetVertexDeclaration(material));
+
+        for (int i = 0; i < infos->m_NumInfos; ++i)
+        {
+            dmParticle::ParticleVertexAttributeInfo& info = infos->m_Infos[i];
+            infos->m_Infos[i].m_NameHash = material_attributes[i].m_NameHash;
+            dmRender::GetMaterialProgramAttributeValues(material, i, (const uint8_t**) &info.m_ValuePtr, &info.m_ValueByteSize);
+        }
+    }
+
```

Comment:
Future improvement: all of these functions are almost identical to comp_sprite, would be nice to merge them since we will follow this pattern later for other components.